### PR TITLE
[SCEV] Prove no-self-wrap from negative power of two step

### PR DESCRIPTION
**Pull request by preames**
We have existing code which reasons about a step evenly dividing the iteration space is a finite loop with a single exit implying no-self-wrap.  The sign of the step doesn't effect this.

### CODE REVIEW

**Review by nikic — COMMENTED** [review]
> Not really a fan of the conditional negate logic here, but it seemed like a bit of overkill to have isKnownToBeAPossibleNegativePowerOfTwo. Any better ideas on how to structure this?

I think it would be okay to add an OrNegative parameter to the function. Computing getNegativeSCEV() may be expensive if it's not just negating a constant.

**Comment by preames** [comment]
> > Not really a fan of the conditional negate logic here, but it seemed like a bit of overkill to have isKnownToBeAPossibleNegativePowerOfTwo. Any better ideas on how to structure this?
> 
> I think it would be okay to add an OrNegative parameter to the function. Computing getNegativeSCEV() may be expensive if it's not just negating a constant.

Will do.  Update pending later today.

File: llvm/lib/Analysis/ScalarEvolution.cpp [hunk 1]
```diff
@@ -9158,7 +9158,8 @@ ScalarEvolution::ExitLimit ScalarEvolution::computeExitLimitFromICmp(
       InnerLHS = ZExt->getOperand();
     if (const SCEVAddRecExpr *AR = dyn_cast<SCEVAddRecExpr>(InnerLHS);
         AR && !AR->hasNoSelfWrap() && AR->getLoop() == L && AR->isAffine() &&
-        isKnownToBeAPowerOfTwo(AR->getStepRecurrence(*this), /*OrZero=*/true)) {
+        isKnownToBeAPowerOfTwo(AR->getStepRecurrence(*this), /*OrZero=*/true,
+                               /*OrNegative*/ true)) {
```
Comment (nikic) [inline llvm/lib/Analysis/ScalarEvolution.cpp:9162]:
```suggestion
                               /*OrNegative=*/true)) {
```
For consistency.

File: llvm/lib/Analysis/ScalarEvolution.cpp [hunk 2]
```diff
@@ -10857,7 +10861,11 @@ bool ScalarEvolution::isKnownToBeAPowerOfTwo(const SCEV *S, bool OrZero) {
   auto *Mul = dyn_cast<SCEVMulExpr>(S);
   if (!Mul)
     return false;
-  return all_of(Mul->operands(), NonRecursive) && (OrZero || isKnownNonZero(S));
+  return all_of(Mul->operands(), NonRecursive) &&
+         (OrZero || isKnownNonZero(S)) &&
+         (!OrNegative || llvm::count_if(Mul->operands(), [this](const SCEV *S) {
```
Comment (nikic) [inline llvm/lib/Analysis/ScalarEvolution.cpp:10866]: I don't get why we need this condition.

If OrNegative=false the previous code is correct. If OrNegative=true, then we're fine with either a positive or negative power of two. So even if we have two negative factors, the result will be a positive power of two (or zero) and still be fine.

Comment (preames) [reply llvm/lib/Analysis/ScalarEvolution.cpp:10866]: You're right.  I was originally thinking of this as power-of-two XOR negative power-of-two, and then generalized slightly, but you're right, this condition is a remnant of the XOR model.  Will kill shortly.

File: llvm/test/Analysis/ScalarEvolution/trip-count-scalable-stride.ll [hunk 3]
```diff
@@ -455,15 +455,16 @@ define void @vscale_countdown_ne(ptr nocapture %A, i32 %n) mustprogress vscale_r
 ; CHECK-NEXT:    %start = sub i32 %n, %vscale
 ; CHECK-NEXT:    --> ((-1 * vscale)<nsw> + %n) U: full-set S: full-set
 ; CHECK-NEXT:    %iv = phi i32 [ %sub, %for.body ], [ %start, %entry ]
-; CHECK-NEXT:    --> {((-1 * vscale)<nsw> + %n),+,(-1 * vscale)<nsw>}<%for.body> U: full-set S: full-set Exits: <<Unknown>> LoopDispositions: { %for.body: Computable }
+; CHECK-NEXT:    --> {((-1 * vscale)<nsw> + %n),+,(-1 * vscale)<nsw>}<%for.body> U: full-set S: full-set Exits: ((vscale * (-1 + (-1 * (((-2 * vscale)<nsw> + %n) /u vscale))<nsw>)<nsw>) + %n) LoopDispositions: { %for.body: Computable }
```
Comment (nikic) [inline llvm/test/Analysis/ScalarEvolution/trip-count-scalable-stride.ll:458]: Drop the TODO above.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**